docs/source/integrations/plugins.md

```diff
@@ -0,0 +1,322 @@
+
+  plugins: [
+    {
+      serverWillStart() {
+        console.log('Server starting!');
+      }
+    }
+  ]
+})
+```
+
+### `requestDidStart`
+
+The `requestDidStart` event fires whenever Apollo Server begins fulfilling a GraphQL request.
+
+This function can optionally return an object that includes functions for responding
+to request lifecycle events that might follow `requestDidStart`.
+
+```js
+const server = new ApolloServer({
+  /* ... other necessary configuration ... */
+
+  plugins: [
+    {
+      requestDidStart(requestContext) {
+
+        /* Within this returned object, define functions that respond
+           to request-specific lifecycle events. */
+        return {
+
+          /* The `parsingDidStart` request lifecycle event fires
+             when parsing begins. The event is scoped within an
+             associated `requestDidStart` server lifecycle event. */
+          parsingDidStart(requestContext) {
+            console.log('Parsing started!')
+          },
+        }
+      }
+    }
+  ],
+})
+```
+
+If your plugin doesn't need to respond to any request lifecycle events, `requestDidStart`
+should not return a value. 
+
+### Request lifecycle events
+
+> If you're using TypeScript to create your plugin, implement the [ `GraphQLRequestListener` interface](https://github.com/apollographql/apollo-server/blob/master/packages/apollo-server-plugin-base/src/index.ts) from the `apollo-server-plugin-base` module to define functions for request lifecycle events.
+
+### `parsingDidStart`
+
+The `parsingDidStart` event fires whenever Apollo Server will parse a GraphQL
+request to create its associated `document` AST.
+
+If Apollo Server receives a request with a query string that matches a _previous_
+request, the associated `document` might already be available in Apollo Server's cache.
+In this case, `parsingDidStart` is _not_ called for the request, because parsing
+does not occur.
+
+```typescript
+parsingDidStart?(
+  requestContext: GraphQLRequestContext<TContext>,
+): (err?: Error) => void | void;
+```
+
+### `validationDidStart`
+
+The `validationDidStart` event fires whenever Apollo Server will validate a 
+request's `document` AST against your GraphQL schema.
+
+Like `parsingDidStart`, this event does _not_ fire if a request's `document` is
+already available in Apollo Server's cache (only successfully validated `document`s are cached by Apollo Server).
+
+The `document` AST is guaranteed to be 
+available at this stage, because parsing must succeed for validation to occur.
+
+```typescript
+validationDidStart?(
+  requestContext: WithRequired<GraphQLRequestContext<TContext>, 'document'>,
+): (err?: ReadonlyArray<Error>) => void | void;
+```
+
+### `didResolveOperation`
+
+The `didResolveOperation` event fires after the `graphql` library successfully
+determines the operation to execute from a request's `document` AST. At this stage,
+both the `operationName` string and `operation` AST are available.
```

Comment:
I think the aforementioned bits about the return value's "purpose" or possible effects (might only be one additional sentence per event) could just be added here and in the other paragraphs respectively.